src/canvas/Canvas.ts

```diff
@@ -1460,193 +1458,126 @@ export class Canvas extends SelectableCanvas {
       !!target &&
       target.selectable &&
-      // if all pre-requisite pass, the target is either something different from the current
-      // activeObject or if an activeSelection already exists
-      // TODO at time of writing why `activeObject.type === 'activeSelection'` matter is unclear.
-      // is a very old condition uncertain if still valid.
-      (activeObject !== target || activeObject.type === 'activeSelection') &&
-      //  make sure `activeObject` and `target` aren't ancestors of each other
-      !target.isDescendantOf(activeObject) &&
-      !activeObject.isDescendantOf(target) &&
+      // group target and active object only if they are different objects
+      // else we try to find a subtarget of `ActiveSelection`
+      (activeObject !== target || isAS) &&
+      //  make sure `activeObject` and `target` aren't ancestors of each other in case `activeObject` is not `ActiveSelection`
+      // if it is then we want to remove `target` from it
+      (isAS ||
+        (!target.isDescendantOf(activeObject) &&
+          !activeObject.isDescendantOf(target))) &&
       //  target accepts selection
-      !target.onSelect({ e: e })
-    );
-  }
-
-  /**
-   * Handles active selection creation for user event
-   * @private
-   * @param {TPointerEvent} e Event object
-   * @param {FabricObject} target
-   */
-  _handleGrouping(e: TPointerEvent, target: FabricObject) {
-    let groupingTarget: FabricObject | undefined = target;
-    // Called always a shouldGroup, meaning that we can trust this._activeObject exists.
-    const activeObject = this._activeObject!;
-    // avoid multi select when shift click on a corner
-    if (activeObject.__corner) {
-      return;
-    }
-    if (groupingTarget === activeObject) {
-      // if it's a group, find target again, using activeGroup objects
-      groupingTarget = this.findTarget(e, true);
-      // if even object is not found or we are on activeObjectCorner, bail out
-      if (!groupingTarget || !groupingTarget.selectable) {
-        return;
-      }
-    }
-    if (activeObject && activeObject.type === 'activeSelection') {
-      this._updateActiveSelection(e, groupingTarget);
-    } else {
-      this._createActiveSelection(e, groupingTarget);
-    }
-  }
-
-  /**
-   * @private
-   */
-  _updateActiveSelection(e: TPointerEvent, target: FabricObject) {
-    const activeSelection = this._activeObject! as ActiveSelection,
-      currentActiveObjects = activeSelection.getObjects();
-    if (target.group === activeSelection) {
-      activeSelection.remove(target);
-      this._hoveredTarget = target;
-      this._hoveredTargets = this.targets.concat();
-      if (activeSelection.size() === 1) {
-        // activate last remaining object
-        this._setActiveObject(activeSelection.item(0) as FabricObject, e);
+      !target.onSelect({ e }) &&
+      // make sure we are not on top of a control
+      !activeObject.__corner
```

Comment:
danger commit

Reply:
This was tricky to validate.
The change seems to effect the following logic because `_shouldGroup` will return false if `activeObject.__corner` which is a change in behavior and since the code is messy it is hard to say if it is safe. However I will prove it doesn't change anything.

So we assume the following:
1. `activeObject.__corner` is defined
2. The original `_shouldGroup` returns true (if it returns `false` there is no change to behavior)

First of all `_handleGrouping` is skipped regardless of this change so we don't need to discuss that. => `activeObject` was not changed

(1) => `target === activeObject` => calculating `_findTargetCorner` is redundant

```diff
    let shouldGroup = false;
    if (this._shouldClearSelection(e, target)) {
      this.discardActiveObject(e);
-    } else if (this._shouldGroup(e, target)) { // true
+    } else if (this._shouldGroup(e, target)) { // false
+    // this block is skipped
-     shouldGroup = true; // remains false
     // may change active object
-      this._handleGrouping(e, target!); // skipped regardless
-      target = this._activeObject; // skipped `_handleGrouping` => no change to `this._activeObject` => `target === this._activeObject`
    }
+   // conclusion: `shouldGroup` is the only thing changed up to here
    // we start a group selector rectangle if
    // selection is enabled
    // and there is no target, or the following 3 condition both apply
    // target is not selectable ( otherwise we selected it )
    // target is not editing
    // target is not already selected ( otherwise we drage )
    if (
      this.selection &&
      (!target ||
        (!target.selectable &&
          // @ts-ignore
          !target.isEditing &&
          target !== this._activeObject))
    ) {
      const p = this.getPointer(e);
      this._groupSelector = {
        ex: p.x,
        ey: p.y,
        top: 0,
        left: 0,
      };
    }


+    if (target) { // true
+      const alreadySelected = target === this._activeObject;  // true
      if (target.selectable && target.activeOn === 'down') {
+        this.setActiveObject(target, e); // redundant since `target === activeObject`
      }
      const corner = target._findTargetCorner(
        this.getPointer(e, true),
        isTouchEvent(e)
+      ); // redundant since `corner === activeObject.__corner` because `target === activeObject`
+      if (target === this._activeObject && (corner || !shouldGroup)) { // corner is defined regardless
+     // conclusion: in this case `shouldGroup` is not used by the code
        this._setupCurrentTransform(e, target, alreadySelected);
        const control = target.controls[corner],
          pointer = this.getPointer(e),
          mouseDownHandler =
            control && control.getMouseDownHandler(e, target, control);
        if (mouseDownHandler) {
          mouseDownHandler(e, this._currentTransform!, pointer.x, pointer.y);
        }
      }
    }
-    const invalidate = shouldRender || shouldGroup; // was true
+    const invalidate = shouldRender || shouldGroup; // now it can be false, this correct since `shouldGroup` made no changes to state
```

https://github.com/fabricjs/fabric.js/blob/e502e1f74dc782c275875323066c7a06c76287be/src/canvas/Canvas.ts#L1098-L1150